File: ios/App/Store/SessionStore.swift
import SwiftUI

@MainActor
final class SessionStore: ObservableObject {
  @Published private(set) var state = AuthState(
    session: nil,
    hydratedUser: nil,
    isHydratingUser: false,
    isSigningIn: false,
    lastError: nil,
    successMessage: nil
  )

  init() {}

  var isAuthenticated: Bool {
    state.session != nil
  }

  var hasHydratedUser: Bool {
    state.hydratedUser != nil
  }

  var session: AuthSession? {
    state.session
  }

  func reportError(_ message: String) {
    state.lastError = message
    state.successMessage = nil
  }

  func reportSuccess(_ message: String) {
    state.lastError = nil
    state.successMessage = message
  }

  func clearStatusMessages() {
    state.lastError = nil
    state.successMessage = nil
  }

  // MARK: - State mutators used by components

  func setSession(_ session: AuthSession?) {
    state.session = session
  }

  func setHydratedUser(_ user: CurrentUserData?) {
    state.hydratedUser = user
  }

  func setHydratingUser(_ isHydrating: Bool) {
    state.isHydratingUser = isHydrating
  }

  func setSigningIn(_ isSigningIn: Bool) {
    state.isSigningIn = isSigningIn
  }

  func resetForSignOut() {
    state.session = nil
    state.hydratedUser = nil
    state.isHydratingUser = false
    state.isSigningIn = false
    clearStatusMessages()
  }
}

File: android/app/src/main/kotlin/com/baseapp/android/store/SessionStore.kt
package com.baseapp.android.store

import com.baseapp.android.auth.AuthSession
import com.baseapp.android.auth.AuthState
import com.baseapp.android.graphql.CurrentUserData
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/** State-only session container (see KOTLIN_ANDROID_STORE_COMPONENT_PATTERN.md). */
class SessionStore {
    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state.asStateFlow()

    val isAuthenticated: Boolean get() = _state.value.session != null
    val hasHydratedUser: Boolean get() = _state.value.hydratedUser != null
    val session: AuthSession? get() = _state.value.session

    fun reportError(message: String) {
        _state.value = _state.value.copy(lastError = message, successMessage = null)
    }

    fun reportSuccess(message: String) {
        _state.value = _state.value.copy(lastError = null, successMessage = message)
    }

    fun clearStatusMessages() {
        _state.value = _state.value.copy(lastError = null, successMessage = null)
    }

    // --- State mutators used by components ---

    fun setSession(session: AuthSession?) {
        _state.value = _state.value.copy(session = session)
    }

    fun setHydratedUser(user: CurrentUserData?) {
        _state.value = _state.value.copy(hydratedUser = user)
    }

    fun setHydratingUser(isHydrating: Boolean) {
        _state.value = _state.value.copy(isHydratingUser = isHydrating)
    }

    fun setSigningIn(isSigningIn: Boolean) {
        _state.value = _state.value.copy(isSigningIn = isSigningIn)
    }

    fun resetForSignOut() {
        _state.value = AuthState()
    }
}
